lib/fizzy/leb128.hpp

```diff
@@ -24,4 +26,28 @@ std::pair<T, const uint8_t*> leb128u_decode(const uint8_t* input)
     throw std::runtime_error("Invalid LEB128 encoding: too many bytes.");
 }
 
+template <typename T>
+std::pair<T, const uint8_t*> leb128s_decode(const uint8_t* input)
+{
+    static_assert(std::numeric_limits<T>::is_signed);
+
+    using T_unsigned = typename std::make_unsigned<T>::type;
+    T_unsigned result = 0;
+    int result_shift = 0;
+
+    for (; result_shift < std::numeric_limits<T_unsigned>::digits; ++input, result_shift += 7)
+    {
+        result |= static_cast<T_unsigned>((static_cast<T_unsigned>(*input) & 0x7F) << result_shift);
+        if ((*input & 0x80) == 0)
+        {
+            // sign extend
```

Comment:
Sign extension has to be done only once in the end. See https://en.wikipedia.org/wiki/LEB128#Decode_signed_integer.

Reply:
It is done only on the last iteration